### PR TITLE
Add checks for infinity and NaN in float check

### PR DESCRIPTION
**Pull request by psafont**
Until now some comparisons between edge-cases in floats were failing.

A test suite for floats has been added in `examples/` to easily compare the before and after.

Edit: Reflecting back on this, there might be reason to keep the standard behaviour, I'm open to suggestion on how to best keep both behaviours.

### CODE REVIEW

File: examples/dune [hunk 1]
```diff
@@ -1,7 +1,7 @@
 (executables
-  (names     bad simple)
+  (names     bad simple floats)
   (libraries alcotest))
 
 (alias
   (name runtest)
-  (deps bad.exe simple.exe))
+  (deps bad.exe simple.exe floats.exe))
```
Comment (emillon) [inline examples/dune]: That's not really related to your changes, but I don't think that this runs the examples. Can you confirm that introducing a runtime error in the test is not caught by `dune runtest`?

File: examples/floats.ml [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+OTHER DEALINGS IN THE SOFTWARE.
+
+For more information, please refer to <http://unlicense.org/>
+*)
+
+let e = epsilon_float
+
+(* The tests *)
```
Comment (emillon) [inline examples/floats.ml]:
```suggestion
```

File: examples/floats.ml [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+let edge_set = [
+  "NaN", `Quick, nan     ;
+  "∞",   `Quick, infinity;
+]
+
+let others_set = [
+  "others", `Quick, others;
+]
+
+(* Run it *)
```
Comment (emillon) [inline examples/floats.ml]:
```suggestion
```

File: src/alcotest.ml [hunk 4]
```diff
@@ -617,7 +619,12 @@ let int64 = testable Fmt.int64 (=)
 
 let int = testable Fmt.int (=)
 
-let float eps = testable Fmt.float (fun x y -> abs_float (x -. y) <= eps)
+let float eps =
+  let same x y = (isnan x && isnan y) ||
```
Comment (emillon) [inline src/alcotest.ml]: what do you think about naming each case to help reading operators? Something like:
```ocaml
let both_are_nan = isnan x && isnan y in
...
both_are_nan || ...
```

File: examples/dune (hunk 1 above)
Comment (psafont) [reply examples/dune]: None of the examples are run on `dune runtest` as there is no action specified for runtest dune file in there.
This line could be added to run simple and floats: `(action (progn (run ./simple.exe) (run ./floats.exe)))`

File: src/alcotest.ml (hunk 4 above)
Comment (psafont) [reply src/alcotest.ml]: Does it bring that much clarity? I'm not convinced

File: examples/dune (hunk 1 above)
Comment (emillon) [reply examples/dune]: I'll fix that separately (the idiomatic way to do this is to turn the executable stanza into a test stanza)

File: src/alcotest.ml (hunk 4 above)
Comment (emillon) [reply src/alcotest.ml]: I think that it would help with understanding operator priority and add an opportunity to explain why the `x = y` case is important.

Comment (psafont) [reply src/alcotest.ml]: Ideally I would use `Float.equal x y || abs_float (x -. y) <= eps`.

In any case I'm not sure how would using `both_are_nan` would clarify the second term is related to infinity values, will a comment do?

Comment (emillon) [reply src/alcotest.ml]: Sure, `Float.equal` works for me!

Comment (psafont) [reply src/alcotest.ml]: Sorry, I forgot to say why that's the ideal solution: it's not available on Ocaml < 4.07. (4.04 to 4.06 seem to be supported)

**Comment by samoht** [comment]
Thanks for your patch, that sounds like a good idea! 

Would you be able to rebase this to master now that #153 is merged?

**Comment by samoht** [comment]
Thanks!
